### ISSUE TITLE
Elmer Documentations and tutorials

### ISSUE THREAD

**Issue by Foadsf** [ElmerCSC/elmerfem#201]
I see that here on [SourceForge](https://sourceforge.net/projects/elmerfem/files/ElmerDocumentation/), there are all Documentations in PDF format archived in single downloadable files. However the source is not provided. Did you use LaTeX or any other MarkUp languages to generate these PDFs? I would appreciate if you could do either of these

 - add a `Documentation` folder on [this repository](https://github.com/ElmerCSC/elmerfem) with the source files
 - create a new repo on the [ElmerCSC](https://github.com/ElmerCSC) GitHub organization with the name `ElmerDocumentation` witht the source files

please remember to update the `.gitignore` file according to the specific MarkUp language you have used. use [this website](https://www.gitignore.io/) to generate templates. 

**P.S.1.** If my vote counts I choose the first option as anyone downloading//installing the software will pose the documentation by default. and somewhere in the help menu of the GUI, we can link to these PDFs.

**P.S.2.** I might attempt to convert the documentation to a more moderns format, for example, the GitHub HTML website using the AsciiDoc MarkUp language. See [this](https://foadsf.github.io/ssct/) for example.

**Comment by Foadsf** [comment]
I see there are a lot of nice videos [here on Elmer's YouTube channel](https://www.youtube.com/user/elmerfem/featured) as well as [this FUNET fttp folder](http://www.nic.funet.fi/pub/sci/physics/elmer/anim/). Ideally we can provide the source code and instructions to all of these clips and make the Elmer's documentation much reacher than what it is now. Also, we may create a White Paper section listing all the studies who have used Elmer or contributed to Elmer in any way (e.g., I tried doing this for Modelica PDE [here](https://github.com/Foadsf/ModelicaPDE) for example).

**Comment by Foadsf** [comment]
I see [here on FUNET fttp](http://www.nic.funet.fi/pub/sci/physics/elmer/), there are folders of [course](http://www.nic.funet.fi/pub/sci/physics/elmer/course/), [slides](http://www.nic.funet.fi/pub/sci/physics/elmer/slides/), and [doc](http://www.nic.funet.fi/pub/sci/physics/elmer/doc/), all with amazing learning resources. I'm actually surprised that these are not better advertised. I would appreciate it if you could add the source to those documents, to the presumed repo/folder as well.

**Comment by Foadsf** [comment]
If this ends up being an independent repo, [these packages](https://repology.org/project/elmer-doc/versions) should be updated accordingly.

**Comment by tehnick** [comment]
> If this ends up being an independent repo, [these packages](https://repology.org/project/elmer-doc/versions) should be updated accordingly.

These packages do not exist anymore:
https://tracker.debian.org/pkg/elmer-doc
https://bugs.debian.org/875442

**Comment by tehnick** [comment]
> Did you use LaTeX or any other MarkUp languages to generate these PDFs?

IIRC all documentation was generated from LaTeX sources. Just look at document properties of new files.

But have in mind that if nothing changed since 2014 this documentation is under non-free license `CC-BY-ND-3.0`, so there are no reasons for re-generating it from sources. Just use finite files provided by developers.

**Comment by Foadsf** [comment]
@tehnick IMHO PDF is not a good way to have documentation. Especially for a program like Elmer which works with code. Ideally, I would prefer to have HTML generated by a lightweight MarkUp language for example AsciiDoc. My goal is to have the source, firstly for the sake of having them here, then convert them. Though Creative Commons licensing is not Free, it is open source, and we can distribute the source code. Am I wrong?

**Comment by tehnick** [comment]
> IMHO PDF is not a good way to have documentation. Especially for a program like Elmer which works with code.

Yes, this is a moot point. For example, we may see this project as software for scientists and engineers. And for them PDF format is one of the most widely used format for file exchange. (See books, articles, presentations, reports, etc.)

> Am I wrong?

Just read short description of this license if you do not want read its full version...

> Ideally, I would prefer to have HTML generated by a lightweight MarkUp language for example AsciiDoc.

LaTeX sources may be easily converted to html pages, markdown files and many other formats. But taking into account the `CC-BY-ND` license this should be done by authors of these files. Though I may be wrong here...

**Comment by Foadsf** [comment]
The `ND` part AFIK means no derivatives. That shouldn't be a problem for distributing the source here on GitHub. But let's see the opinion of @raback . Changing licensing is also a possibility.

**Comment by raback** [comment]
The files have indeed been written in Latex. First parts of the documentation comes from the 90's and then that was the only tool to make pretty formulas. I know that nowadays it is possible to have pretty formulas also using different markup languages but this is not reason enough to move the documentation completely to new platform. There is almost 800 pages of documentation and in a project of that size Latex is still pretty good. In recent years there has been modifications mainly to Models manual and lesser degree Solver manual. So even though most of the documentation is quite static it is still edited regularly.

When the documentation was published it was about balancing the freedom and copyrights. As the code is still mainly (but by no means only) developed by CSC and there were concerns that giving the LaTex source out would make some spin-offs too easy. I think that this is nowadays less of a concern. Still, dropping the "ND" and releasing all the Latex files would require a formal decision. 

I don't think Elmer is alone among open source code which still limits the access to documentation. On the other hand I see there would be clear benefits in having the documentation browsable.

**Comment by Foadsf** [comment]
Dear @raback , 

Thanks for the reply. Although LaTeX/PDF is an excellent way for publishing academic literature, it is not necessarily the best option for a FLOSS tool that aspires to be used by everyone. Why PDF is not useful:

 - For a code-based software like Elmer, in contrast to GUI based ones such as ANSYS and COMSOL, the ability to format code nicely embedded into the documentation and the ability to modularize the code parts is essential. Also, code embedded in LaTeX/PDFs are not easy to copy past for users, usually needs to be done line by line, and formatting often is not copied correctly. 
 - PDFs are not easily indexed by search engines and are, in general, less searchable than other MarkUp formats such as HTML.
 - PDFs being binary files are not very easy to do version control on

While LaTeX is a very widely used MarkUp language, the software community seems to be moving away to the more lightweight languages such as MarkDown, AsciiDoc, reStructuredText ... which can be rendered in realtime, they can embed pretty mathematical equations, they are rendered by GitHub, GitLab, and other modern repository management systems. The process of conversion can be done gradually, and the PDF format can coexist side by side. I'm personally a fan of AsciiDoc, as it seems to be one of the most capable ones, and it has a great team and community supporting it.

**Comment by raback** [comment]
Yes, I agree that today Markdown would be preferable particularly for the community. Certainly if we would start from scratch the choices would be very different in many respects. I guess the question is whether there would be any reasonable way to semi-automatically translate the Latex documentation to some markdown language. I guess basic latex is easy to translate to anything but formulas will be the bottle-neck (and maybe citations and cross references). 

The order of formulas is probably in range of ~1000. If we would need one minute to fix the formula for markdown language just that would be 17 hours of crazy work. So the converter has to be so solid that it can deal with >>90% of the formulas. Long ago Latex2html worked pretty ok but it the deficiency that every formula was a picture. To my understanding only Latex can properly interpret Latex.

Another strategy could be to take just part of the documentation. For example, ElmerGUI tutorials would be easy as they don't include almost any formulas. Also there the community could probably have a stronger role than in writing, say, the ElmerSolver library functionality. Also the tutorials require other files which are easily bundled.

**Comment by kinnala** [comment]
Is Models Manual the only document where the solver keywords are defined? I see the benefit of having those in a machine-readable form, e.g., to create better editor support for *.sif-files.

**Comment by kinnala** [comment]
Seems that the easiest way to get a list of solver keywords is to apply `ripgrep` on `elmerfem/fem/src/modules`:
```
conda-shell-chrootenv:tom@tunkki:~/src/elmerfem/elmer/fem/src/modules$ rg GetLogical
...
contrib/ShellMultiSolver/ShellMultiSolver.F90
257:     LargeDeflection =  GetLogical( SolverParams, 'Large Deflection', GotIt )
277:       StressComputation = GetLogical( SolverParams, 'StressComputation', GotIt )
279:           GetLogical( SolverParams, 'Calculate Stresses', GotIt )
284:     StabilityAnalysis = GetLogical( SolverParams, 'Stability Analysis', GotIt )
309:     SimpleObject = GetLogical( SolverParams, 'Object Of Simple Regular Shape', GotIt )
310:     TopSideStress = GetLogical( SolverParams, 'Top Side Stress', GotIt )
311:     BottomSideStress = GetLogical( SolverParams, 'Bottom Side Stress', GotIt )
318:     TotalStress = GetLogical( SolverParams, 'Compute Total Stress', GotIt )
319:     MembraneStress = GetLogical( SolverParams, 'Compute Membrane Stress', GotIt )
320:     BendingStress = GetLogical( SolverParams, 'Compute Bending Stress', GotIt )
332:     MembraneOnly = GetLogical( SolverParams, 'Membrane Only', GotIt )	
333:     MembraneOnlyNRM = GetLogical( SolverParams, 'Membrane Only No RM', GotIt )
335:     UseDKTtriangle = GetLogical( SolverParams, 'Use DKT Triangle', GotIt )
336:     UseRDKTtriangle = GetLogical(SolverParams, 'Use RDKT Triangle', GotIt )
337:     UseSMITCelement = GetLogical( SolverParams, 'Use SMITC Element', GotIt )
338:     UseMITC3element = GetLogical( SolverParams, 'Use MITC3 Element', GotIt )
339:     UseRMITC3element = GetLogical( SolverParams, 'Use RMITC3 Element', GotIt )
543:       StressComputation = GetLogical( SolverParams, 'Stress Computation', GotIt )
545:           GetLogical( SolverParams, 'Calculate Stresses', GotIt )
866:        IF( ListGetLogical( SolverParams,'Compute Membrane Stress',GotIt) ) THEN
883:        ELSE IF( ListGetLogical( SolverParams,'Compute Bending Stress',GotIt) ) THEN  
898:        ELSEIF( ListGetLogical( SolverParams, 'Compute Total Stress', GotIt ) ) THEN
1223:      WeldSimu = GetLogical( BodyForce, 'Weld Distortion Simulation', GotIt )
1226:      ThStrain = GetLogical( BodyForce, 'Calculate Thermal Strain', GotIt )
1233:      AnisoPlane = GetLogical( BodyForce, 'Use Relaxed AnisoPlane Testing', GotIt )
2235:       WeldSimu = GetLogical( BodyForce, 'Weld Distortion Simulation', GotIt )
2237:       ThStrain = GetLogical( BodyForce, 'Calculate Thermal Strain', GotIt )
2243:       AnisoPlane = GetLogical( BodyForce, 'Use Relaxed AnisoPlane Testing', GotIt )
3143:       WeldSimu = GetLogical(BodyForce, 'Weld Distortion Simulation', Found1 )
```
I think this is basically a complete list of allowed keywords?

**Comment by ettaka** [comment]
Hi Tom,

in principle solver keywords are here:
https://github.com/ElmerCSC/elmerfem/blob/devel/fem/src/SOLVER.KEYWORDS
But it is not always up to date. For example I am quilty of adding features
without appending that file which kind of sucks but I never remember!

which editor are you using?
I have a small plugin for vim/neovim if you are interested:
https://github.com/ettaka/vim-elmer

Cheers,

Eelis

pe 16. lokak. 2020 klo 19.37 Tom Gustafsson (notifications@github.com)
kirjoitti:

**Comment by kinnala** [comment]
@ettaka Thanks for the input!

In fact I'm creating a Python module which attempts to help in the creation of *.sif files and would like to include some basic validation of the user input. I already wrote a mesh converter from NumPy arrays to Elmer format and am now working on a set of functions for defining the solvers. Eventually it should also have a function for launching Elmer with the mesh and the solvers defined by the user.

I'm doing this mostly to force myself to learn a bit of Elmer but if it becomes something more than a prototype I'll release it.

**Comment by ettaka** [comment]
Sounds excellent! Keep us posted, very interesting!

pe 16. lokak. 2020 klo 19.51 Tom Gustafsson (notifications@github.com)
kirjoitti:

**Comment by raback** [comment]
The known keywords can be found in $ELMERSRC/fem/src/SOLVER.KEYWORDS

Additionally, I sometimes use compilation flags "DEVEL_KEYWORDMISSES" and "DEVEL_LISTCOUNTER". The first one writes missing keywords to file and the 2nd one counts how many times certain keywords are accessed.

**Comment by raback** [comment]
juhanikataja had also an idea to have a more rigorous data type that would be used to ensure correctness of the sif file. Maybe he could share his ideas. I think the challenge is that having this full-proof is not that easy.

**Comment by raback** [comment]
ElmerGUI manual & tutorials are now available on GitHub
https://github.com/ElmerCSC/elmerfem-manuals
Unfortunately this does not yet apply to Elmer Models Manual & ElmerSolver Manual.

**Comment by kinnala** [comment]
Where should we report issues on Elmer Models Manual and ElmerSolver Manual?

**Comment by richb2k** [comment]
Would it be possible to add some of the other documents, other than the Model or Solver Manuals, to elmerfem-manuals on github?  Such as: ElmerGridManual, ElmerOverview, and MATCManual?  Thanks.

**Comment by raback** [comment]
Elmergrid, overview and matc are now on the public side.

**Comment by raback** [comment]
You may request access to models and solver as well. At the moment they are not in a open repo.

**Comment by richb2k** [comment]
Hello Peter,   Thank you for moving over those three document.  Rich.
